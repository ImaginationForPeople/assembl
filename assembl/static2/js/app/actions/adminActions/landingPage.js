// @flow
import * as actionTypes from '../actionTypes';

type toggleLandingPageModuleAction = string => actionTypes.toggleLandingPageModule;
export const toggleLandingPageModule: toggleLandingPageModuleAction = moduleTypeIdentifier => ({
  moduleTypeIdentifier: moduleTypeIdentifier,
  type: actionTypes.TOGGLE_LANDING_PAGE_MODULE
});

type updateLandingPageModulesAction = actionTypes.LandingPageModules => actionTypes.UpdateLandingPageModules;
export const updateLandingPageModules: updateLandingPageModulesAction = modules => ({
  modules: modules,
  type: actionTypes.UPDATE_LANDING_PAGE_MODULES
});