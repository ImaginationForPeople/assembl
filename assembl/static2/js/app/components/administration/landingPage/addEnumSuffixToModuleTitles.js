// @flow
import countBy from 'lodash/countBy';
import get from 'lodash/get';

export const addEnumSuffixToModuleTitles = (modules: Array<LandingPageModule>): Array<LandingPageModule> => {
  // Add a suffix to the title of the module if this module appears more than one time.
  // This suffix is the number of times this module appeared in the array.
  const moduleTypeTitles = modules.map(module => ({ title: module.moduleType.title }));
  const titleCounts = countBy(moduleTypeTitles, 'title');
  const duplicatesCurrentIndex = {};
  return modules.map((module) => {
    const { title } = module.moduleType;
    if (title && titleCounts[title] > 1) {
      duplicatesCurrentIndex[title] = get(duplicatesCurrentIndex, title, 0) + 1;
      return { ...module, moduleType: { ...module.moduleType, title: title && `${title} ${duplicatesCurrentIndex[title]}` } };
    }
    return module;
  });
};