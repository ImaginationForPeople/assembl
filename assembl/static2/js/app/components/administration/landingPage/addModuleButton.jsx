// @flow
import React from 'react';
import { Button } from 'react-bootstrap';
import { Translate } from 'react-redux-i18n';
import { closeModal, displayModal } from '../../../utils/utilityManager';

type Props = {
  buttonTitleTranslationKey: string,
  createModule: Function,
  disabled?: boolean
};

const AddModuleButton = ({ buttonTitleTranslationKey, createModule, disabled }: Props) => {
  const displayConfirmationModal = () => {
    const body = <Translate value="administration.landingPage.manageModules.confirmationModal" />;
    const footer = [
      <Button key="cancel" id="cancel-deleting-button" onClick={closeModal} className="button-cancel button-dark">
        <Translate value="cancel" />
      </Button>,
      <Button
        key="add"
        id="confirm-add-tm-button"
        onClick={() => {
          createModule();
          closeModal();
        }}
        className="button-submit button-dark"
      >
        <Translate value="validate" />
      </Button>
    ];
    const includeFooter = true;
    return displayModal(null, body, includeFooter, footer);
  };
  return (
    <Button className="button-submit button-dark" onClick={disabled ? () => false : displayConfirmationModal} disabled={disabled}>
      <Translate value={`administration.landingPage.manageModules.${buttonTitleTranslationKey}`} />
    </Button>
  );
};

AddModuleButton.defaultProps = {
  disabled: false
};

export default AddModuleButton;