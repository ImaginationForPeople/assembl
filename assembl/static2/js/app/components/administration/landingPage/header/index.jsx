// @flow
import React from 'react';
import { connect } from 'react-redux';
import { type ApolloClient, graphql, compose, withApollo } from 'react-apollo';
import { I18n } from 'react-redux-i18n';
import { Field } from 'react-final-form';
import setFieldTouched from 'final-form-set-field-touched';
import SectionTitle from '../../../administration/sectionTitle';
import Helper from '../../../common/helper';
import { load, postLoadFormat } from './load';
import { save, createMutationsPromises } from './save';
import LoadSaveReinitializeForm from '../../../form/LoadSaveReinitializeForm';
import FileUploaderFieldAdapter from '../../../form/fileUploaderFieldAdapter';
import DatePickerFieldAdapter from '../../../form/datePickerFieldAdapter';
import MultilingualTextFieldAdapter from '../../../form/multilingualTextFieldAdapter';
import MultilingualRichTextFieldAdapter from '../../../form/multilingualRichTextFieldAdapter';
import AdminForm from '../../../form/adminForm';
import Loader from '../../../common/loader';
import { validStartDate, validEndDate, validateDatePicker } from './validate';
import DiscussionQuery from '../../../../graphql/DiscussionQuery.graphql';
import { convertISO8601StringToDate } from '../../../../utils/globalFunctions';
import manageErrorAndLoading from '../../../common/manageErrorAndLoading';

type Props = {
  client: ApolloClient,
  editLocale: string,
  data: {
    discussion: {
      startDate: string,
      endDate: string
    }
  }
};

type State = {
  startDate: moment$Moment | null,
  endDate: moment$Moment | null,
  startDateConflict: boolean,
  endDateConflict: boolean
};

const loading = <Loader />;

export class DumbCustomizeHeader extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    const { startDate, endDate } = props.data.discussion;
    this.state = {
      startDate: startDate ? convertISO8601StringToDate(startDate) : null,
      endDate: endDate ? convertISO8601StringToDate(endDate) : null,
      startDateConflict: false,
      endDateConflict: false
    };
  }

  onStartChange = (e: moment$Moment) => {
    const { endDate } = this.state;
    let startDateConflict = false;
    if (!validStartDate(e, endDate)) {
      startDateConflict = true;
    } else {
      startDateConflict = false;
    }
    this.setState(() => ({
      startDate: e,
      startDateConflict: startDateConflict
    }));
  };

  onEndChange = (e: moment$Moment) => {
    const { startDate } = this.state;
    let endDateConflict = false;
    if (!validEndDate(startDate, e)) {
      endDateConflict = true;
    } else {
      endDateConflict = false;
    }
    this.setState(() => ({
      endDate: e,
      endDateConflict: endDateConflict
    }));
  };

  render() {
    const { client, editLocale } = this.props;
    const { startDateConflict, endDateConflict } = this.state;
    return (
      <div className="admin-box">
        <SectionTitle
          title={I18n.t('administration.landingPage.header.title')}
          annotation={I18n.t('administration.annotation')}
        />
        <LoadSaveReinitializeForm
          load={fetchPolicy => load(client, editLocale, fetchPolicy)}
          loading={loading}
          postLoadFormat={postLoadFormat}
          createMutationsPromises={createMutationsPromises(client)}
          save={save}
          mutators={{ setFieldTouched: setFieldTouched }}
          validate={validateDatePicker}
          render={({ handleSubmit, pristine, submitting, form }) => (
            <div className="admin-content">
              <AdminForm
                handleSubmit={handleSubmit}
                pristine={pristine || startDateConflict || endDateConflict}
                submitting={submitting}
              >
                <div className="form-container">
                  <Helper
                    classname="admin-paragraph"
                    label={I18n.t('administration.landingPage.header.helper')}
                    helperUrl={'/static2/img/helpers/landing_page_admin/header.png'}
                    helperText={I18n.t('administration.helpers.landingPage.header')}
                  />
                  <Field
                    editLocale={editLocale}
                    name="headerTitle"
                    component={MultilingualTextFieldAdapter}
                    label={I18n.t('administration.landingPage.header.titleLabel')}
                    required
                  />
                  <Field
                    editLocale={editLocale}
                    withAttachmentButton={false}
                    name="headerSubtitle"
                    component={MultilingualRichTextFieldAdapter}
                    label={I18n.t('administration.landingPage.header.subtitleLabel')}
                    required
                  />
                  <Field
                    editLocale={editLocale}
                    name="headerButtonLabel"
                    component={MultilingualTextFieldAdapter}
                    label={I18n.t('administration.landingPage.header.buttonLabel')}
                    required={false}
                  />
                  <Field
                    name="headerImage"
                    component={FileUploaderFieldAdapter}
                    label={I18n.t('administration.landingPage.header.headerImage')}
                  />
                  <div className="flex">
                    <Field
                      name="headerLogoImage"
                      component={FileUploaderFieldAdapter}
                      label={I18n.t('administration.landingPage.header.logoDescription')}
                    />
                    {/* Passing an empty class to the popOver cancels the default styling */}
                    <Helper helperText={I18n.t('administration.landingPage.header.logoHelper')} popOverClass=" " />
                  </div>
                  <div className="section-description">{I18n.t('administration.landingPage.header.dateDescription')}</div>
                  <Field
                    name="headerStartDate"
                    component={DatePickerFieldAdapter}
                    picker={{ pickerType: I18n.t('administration.landingPage.header.startDate') }}
                    editLocale={editLocale}
                    placeHolder={I18n.t('administration.landingPage.header.timePlaceholder')}
                    showTime={false}
                    hasConflictingDate={this.state.startDateConflict}
                    onDateChange={this.onStartChange}
                    form={form}
                    dateFormat="LL"
                  />
                  <Field
                    name="headerEndDate"
                    component={DatePickerFieldAdapter}
                    picker={{ pickerType: I18n.t('administration.landingPage.header.endDate') }}
                    editLocale={editLocale}
                    placeHolder={I18n.t('administration.landingPage.header.timePlaceholder')}
                    showTime={false}
                    hasConflictingDate={this.state.endDateConflict}
                    onDateChange={this.onEndChange}
                    form={form}
                    dateFormat="LL"
                  />
                </div>
              </AdminForm>
            </div>
          )}
        />
      </div>
    );
  }
}

const mapStateToProps = ({ admin: { editLocale } }) => ({
  editLocale: editLocale
});

export default compose(
  connect(mapStateToProps),
  graphql(DiscussionQuery, {
    options: ({ editLocale }) => ({
      variables: {
        lang: editLocale,
        nextView: null
      }
    })
  }),
  manageErrorAndLoading({ displayLoader: true }),
  withApollo
)(DumbCustomizeHeader);