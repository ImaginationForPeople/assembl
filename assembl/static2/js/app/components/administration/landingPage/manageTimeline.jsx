// @flow
import React from 'react';
import { connect } from 'react-redux';
import { I18n } from 'react-redux-i18n';
import SectionTitle from '../../administration/sectionTitle';
import PhaseForm from './phaseForm';
import Helper from '../../common/helper';

type Props = {
  timelineModuleId: string, // eslint-disable-line react/no-unused-prop-types
  discussionPhaseIds: Array<string>,
  editLocale: string
};

const DumbManageTimeline = ({ discussionPhaseIds, editLocale }: Props) => (
  <div className="admin-box">
    <SectionTitle title={I18n.t('administration.landingPage.timeline.title')} annotation={I18n.t('administration.annotation')} />
    <div className="admin-content">
      <div className="form-container">
        <div className="margin-l">
          <Helper
            label={I18n.t('administration.landingPage.timeline.phaseSection')}
            helperUrl="/static2/img/helpers/landing_page_admin/timeline_phase.png"
            helperText={I18n.t('administration.helpers.timelinePhases')}
            classname="title"
          />
          {discussionPhaseIds.map((phaseId, index) => (
            <PhaseForm key={`phase-${index}`} phaseId={phaseId} editLocale={editLocale} index={index} />
          ))}
        </div>
      </div>
    </div>
  </div>
);

const mapStateToProps = (state, { editLocale }) => {
  const { phasesById } = state.admin.timeline;
  return {
    discussionPhaseIds: phasesById
      .filter(phase => !phase.get('_toDelete'))
      .keySeq()
      .toJS(),
    editLocale: editLocale
  };
};

export default connect(mapStateToProps)(DumbManageTimeline);