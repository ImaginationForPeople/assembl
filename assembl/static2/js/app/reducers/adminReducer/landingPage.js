// @flow
import { fromJS, Map } from 'immutable';
import { combineReducers } from 'redux';
import type ReduxAction from 'redux';

import { type Action, TOGGLE_LANDING_PAGE_MODULE, UPDATE_LANDING_PAGE_MODULES } from '../../actions/actionTypes';

type ModulesHasChangedReducer = (boolean, ReduxAction<Action>) => boolean;
export const modulesHasChanged: ModulesHasChangedReducer = (state = false, action) => {
  switch (action.type) {
  case TOGGLE_LANDING_PAGE_MODULE:
    return true;
  case UPDATE_LANDING_PAGE_MODULES:
    return false;
  default:
    return state;
  }
};

const initialState = Map();
type ModulesReducer = (Map<string, Map>, ReduxAction<Action>) => Map<string, Map>;
export const modules: ModulesReducer = (state = initialState, action) => {
  switch (action.type) {
  case TOGGLE_LANDING_PAGE_MODULE: {
    const moduleType = action.moduleTypeIdentifier;
    return state.updateIn([moduleType, 'enabled'], v => !v);
  }
  case UPDATE_LANDING_PAGE_MODULES: {
    let newState = Map();
    action.modules.forEach((module) => {
      newState = newState.set(module.moduleType.identifier, fromJS(module));
    });
    return newState;
  }
  default:
    return state;
  }
};

export type LandingPageReducer = {
  modulesHasChanged: ModulesHasChangedReducer,
  modules: ModulesReducer
};
const reducers: LandingPageReducer = {
  modulesHasChanged: modulesHasChanged,
  modules: modules
};
export default combineReducers(reducers);