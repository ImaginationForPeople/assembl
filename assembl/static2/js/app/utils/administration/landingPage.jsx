/* Plugin for landing page administration */
import LandingPageModules from '../../graphql/LandingPageModules.graphql';

const landingPagePlugin = {
  graphqlQuery: LandingPageModules,
  loading: 'landingPageLoading',
  hasErrors: 'landingPageHasErrors',
  queryOptions: ({ i18n }) => ({
    variables: { lang: i18n.locale }
  }),
  dataToProps: ({ data }) => {
    if (data.loading) {
      return {
        landingPageLoading: true,
        landingPageModules: []
      };
    }

    if (data.error) {
      return {
        landingPageHasErrors: true,
        landingPageLoading: false,
        landingPageModules: []
      };
    }

    return {
      landingPageLoading: data.loading,
      landingPageHasErrors: data.error,
      refetchLandingPageModules: data.refetch,
      landingPageModules: data.landingPageModules
    };
  }
};

export default landingPagePlugin;