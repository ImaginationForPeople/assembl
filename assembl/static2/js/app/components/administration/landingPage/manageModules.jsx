// @flow
import React from 'react';
import { connect } from 'react-redux';
import { I18n, Translate } from 'react-redux-i18n';
import type { List, Map } from 'immutable';

import ModulesPreview from './modulesPreview';
import SectionTitle from '../../administration/sectionTitle';
import SelectModulesForm from './selectModulesForm';
import { toggleLandingPageModule } from '../../../actions/adminActions/landingPage';

type Props = {
  enabledModulesInOrder: List<Map>,
  locale: string,
  modulesByIdentifier: Map<string, Map>,
  toggleModule: Function
};

export const DumbManageModules = ({ enabledModulesInOrder, locale, modulesByIdentifier, toggleModule }: Props) => (
  <div className="admin-box">
    <SectionTitle
      title={I18n.t('administration.landingPage.manageModules.title')}
      annotation={I18n.t('administration.annotation')}
    />
    <div className="admin-content form-container">
      <Translate value="administration.landingPage.manageModules.helper" />
      <div className="two-columns-admin">
        <div className="column-left">
          <SelectModulesForm lang={locale} modulesByIdentifier={modulesByIdentifier} toggleModule={toggleModule} />
        </div>
        <div className="column-right">
          <ModulesPreview modules={enabledModulesInOrder} />
        </div>
      </div>
    </div>
  </div>
);

const mapStateToProps = (state) => {
  const { enabledModulesInOrder, modulesByIdentifier } = state.admin.landingPage;
  return {
    enabledModulesInOrder: enabledModulesInOrder.map(identifier => modulesByIdentifier.get(identifier)),
    modulesByIdentifier: modulesByIdentifier
  };
};

const mapDispatchToProps = dispatch => ({
  toggleModule: identifier => dispatch(toggleLandingPageModule(identifier))
});

export default connect(mapStateToProps, mapDispatchToProps)(DumbManageModules);