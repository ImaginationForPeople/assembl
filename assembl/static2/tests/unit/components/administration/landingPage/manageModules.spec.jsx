import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { DumbManageModules } from '../../../../../js/app/components/administration/landingPage/manageModules';
import { enabledModulesInOrder, modulesByIdentifier } from './fakeData';

describe('ManageModules component', () => {
  it('should render a form to manage the landing page modules', () => {
    const props = {
      enabledModulesInOrder: enabledModulesInOrder,
      modulesByIdentifier: modulesByIdentifier
    };
    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<DumbManageModules {...props} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});