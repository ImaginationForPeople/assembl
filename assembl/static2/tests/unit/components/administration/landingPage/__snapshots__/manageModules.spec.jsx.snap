// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageModules component should render a form to manage the landing page modules 1`] = `
<div
  className="admin-box"
>
  <SectionTitle
    annotation="Annotation"
    title="Title"
  />
  <div
    className="admin-content form-container"
  >
    <Translate
      value="administration.landingPage.manageModules.helper"
    />
    <div
      className="two-columns-admin"
    >
      <div
        className="column-left"
      >
        <Apollo(Component)
          lang={undefined}
          modulesByIdentifier={
            Immutable.Map {
              HEADER: Immutable.Map {
                enabled: true,
                moduleType: Immutable.Map {
                  identifier: "HEADER",
                  title: "Header",
                },
                order: 1,
              },
              INTRODUCTION: Immutable.Map {
                enabled: false,
                moduleType: Immutable.Map {
                  identifier: "INTRODUCTION",
                  title: "Introduction",
                },
                order: 2,
              },
              VIDEO: Immutable.Map {
                enabled: true,
                moduleType: Immutable.Map {
                  identifier: "VIDEO",
                  title: "Video",
                },
                order: 5,
              },
              FOOTER: Immutable.Map {
                enabled: true,
                moduleType: Immutable.Map {
                  identifier: "FOOTER",
                  title: "Footer",
                },
                order: 99,
              },
            }
          }
          toggleModule={undefined}
        />
      </div>
      <div
        className="column-right"
      >
        <ModulesPreview
          modules={
            Immutable.List [
              Immutable.Map {
                enabled: true,
                moduleType: Immutable.Map {
                  identifier: "HEADER",
                  title: "Header",
                },
                order: 1,
              },
              Immutable.Map {
                enabled: true,
                moduleType: Immutable.Map {
                  identifier: "VIDEO",
                  title: "Video",
                },
                order: 5,
              },
              Immutable.Map {
                enabled: true,
                moduleType: Immutable.Map {
                  identifier: "FOOTER",
                  title: "Footer",
                },
                order: 99,
              },
            ]
          }
        />
      </div>
    </div>
  </div>
</div>
`;
